src/Microsoft.Framework.Runtime/NuGet/Authoring/ManifestReader.cs

```diff
@@ -152,33 +148,36 @@ private static List<ManifestReferenceSet> ReadReferenceSets(XElement referencesE
             }
 
             var references = ReadReference(referencesElement, throwIfEmpty: false);
-            if (references.Count > 0)
+            if (references.Any())
             {
                 // old format, <reference> is direct child of <references>
-                var referenceSet = new ManifestReferenceSet
-                {
-                    References = references
-                };
-                return new List<ManifestReferenceSet> { referenceSet };
+                var referenceSet = new PackageReferenceSet(references);
+                return new List<PackageReferenceSet> { referenceSet };
             }
             else
             {
                 var groups = referencesElement.ElementsNoNamespace("group");
-                return (from element in groups
-                        select new ManifestReferenceSet
-                        {
-                            TargetFramework = element.GetOptionalAttributeValue("targetFramework").SafeTrim(),
-                            References = ReadReference(element, throwIfEmpty: true)
-                        }).ToList();
+                return groups.Select(element =>
+                {
+                    var framework = element.GetOptionalAttributeValue("targetFramework")?.Trim();
+                    if (framework != null)
+                    {
+                        return new PackageReferenceSet(VersionUtility.ParseFrameworkName(framework), ReadReference(element, throwIfEmpty: true));
+                    }
+                    else
+                    {
+                        return new PackageReferenceSet(ReadReference(element, throwIfEmpty: true));
+                    }
+                }).ToList();
             }
         }
 
-        public static List<ManifestReference> ReadReference(XElement referenceElement, bool throwIfEmpty)
+        public static IEnumerable<string> ReadReference(XElement referenceElement, bool throwIfEmpty)
```

Comment:
So why are we using `ElementsNoNamespace` almost everywhere? It's semantically incorrect because it strips the namespace and it may capture elements that should not be captured (they were put to different namespace to not be captured). It also does not save you any typing because if you don't expect your elements to belong to a namespace `referenceElement.Element("reference")` is shorter than `referenceElement.ElementsNoNamespace("reference")` (if you do `referenceElement.Element(ns + "reference")` is still shorter). On top of that if you use just the `Element()` method you can use explicit conversion operators.